bundles/org.openhab.binding.pulseaudio/src/main/java/org/openhab/binding/pulseaudio/internal/PulseAudioAudioSink.java

```diff
@@ -74,69 +78,123 @@ public void process(@Nullable AudioStream audioStream)
-                                scheduler.schedule(() -> soundPlayed.complete(null), timeToWait, TimeUnit.MILLISECONDS);
-                                return soundPlayed;
-                            } else {
-                                return CompletableFuture.completedFuture(null);
-                            }
-                        } else {
-                            // We have a second method available to guess the duration, and it is during transfer
-                            Long timeStampEnd = audioSinkUtils.transferAndAnalyzeLength(normalizedPCMStream,
-                                    clientSocketLocal.getOutputStream(), TARGET_FORMAT);
-                            CompletableFuture<@Nullable Void> soundPlayed = new CompletableFuture<>();
-                            if (timeStampEnd != null) {
-                                long now = System.nanoTime();
-                                long timeToWait = timeStampEnd - now;
-                                if (timeToWait > 0) {
-                                    scheduler.schedule(() -> soundPlayed.complete(null), timeToWait,
-                                            TimeUnit.NANOSECONDS);
-                                }
-                                return soundPlayed;
-                            } else {
-                                return CompletableFuture.completedFuture(null);
+                    AudioWaveUtils.removeFMT(pcmSignedStream);
+                } catch (IOException e) {
+                    logger.warn("IOException trying to remove wav header: {}", e.getMessage());
+                }
+            }
+            if (pcmSignedStream instanceof SizeableAudioStream sizeableAudioStream) {
+                long length = sizeableAudioStream.length();
+                var audioFormat = pcmSignedStream.getFormat();
+                long byteRate = (Objects.requireNonNull(audioFormat.getBitDepth()) / 8)
+                        * Objects.requireNonNull(audioFormat.getFrequency())
+                        * Objects.requireNonNull(audioFormat.getChannels());
+                float durationInSeconds = (float) length / byteRate;
+                duration = Math.round(durationInSeconds * 1000);
+                logger.debug("Duration of input stream : {}", duration);
+            }
+        } else {
+            try {
+                var convertedInputStream = new ConvertedInputStream(audioStream);
+                duration = convertedInputStream.getDuration();
+                pcmSignedStream = convertedInputStream;
+            } catch (UnsupportedAudioFileException | UnsupportedAudioFormatException | IOException e) {
+                return CompletableFuture.failedFuture(new UnsupportedAudioFormatException(
+                        "Cannot send sound to the pulseaudio sink", audioStream.getFormat(), e));
+            }
+        }
+        Optional<SimpleProtocolTCPModule> spModule = Optional.empty();
+        try (pcmSignedStream) {
```

Comment:
Isn't there an issue when playing PipedAudioStream ?

With a try-with, the Closeable resource is always closed at the end of the block.
But, in the case of a PipedAudioStream, the sound is played asynchronously, in another thread. So the try-with block will end, without considering the asyncStream.
In this case, wouldn't the stream be closed to soon ?

Reply:
Yes, it was wrong, I have made some refactor/fixes in the processAndComplete sink method. I think that now the audio stream and the module are closed/released in the correct moment.